src/vs/workbench/contrib/chat/browser/chatListRenderer.ts

```diff
@@ -399,29 +399,92 @@ export class ChatListItemRenderer extends Disposable implements ITreeRenderer<Ch
+			element.renderData ??= { renderedParts: [] };
+			const renderedParts = element.renderData.renderedParts;
+			const wordCountResults: IWordCountResult[] = [];
+			const partsToRender: IChatResponseRenderData['renderedParts'] = [];
+
+			currentResponseData.forEach((part, index) => {
+				const renderedPart = renderedParts[index];
+				// Is this part completely new?
+				if (!renderedPart) {
+					if ('label' in part) {
+						partsToRender[index] = part;
+					} else {
+						const wordCountResult = this.getDataForProgressiveRender(element, part);
+						if (wordCountResult !== undefined) {
+							partsToRender[index] = {
+								renderedWordCount: wordCountResult.actualWordCount,
+								lastRenderTime: Date.now(),
+								isFullyRendered: wordCountResult.isFullString,
+							};
+							wordCountResults[index] = wordCountResult;
+						}
+					}
+				}
+
+				// Did this part's type change from being a string to tree data?
+				else if ('label' in part && !('label' in renderedPart)) {
+					partsToRender[index] = part;
+				}
+
+				// Did this part's content change?
+				else if (!('label' in part) && !('label' in renderedPart) && !renderedPart.isFullyRendered) {
+					const wordCountResult = this.getDataForProgressiveRender(element, part, renderedPart);
+					if (wordCountResult !== undefined) {
+						partsToRender[index] = {
+							renderedWordCount: wordCountResult.actualWordCount,
+							lastRenderTime: Date.now(),
+							isFullyRendered: wordCountResult.isFullString,
+						};
+						wordCountResults[index] = wordCountResult;
+					}
+				}
+			});
+
+			isFullyRendered = partsToRender.length === 0;
+
 			if (isFullyRendered && element.isComplete) {
 				// Response is done and content is rendered, so do a normal render
 				this.traceLayout('runProgressiveRender', `end progressive render, index=${index} and clearing renderData, response is complete, index=${index}`);
 				element.renderData = undefined;
 				disposables.clear();
 				this.basicRenderElement(element.response.value, element, index, templateData);
-			} else if (renderValue) {
-				element.renderData = {
-					renderedWordCount: renderValue.actualWordCount,
-					lastRenderTime: Date.now(),
-					isFullyRendered: renderValue.isFullString
-				};
-
-				const plusCursor = (renderValue.value.match(/```\s*$/) ?
-					renderValue.value + '\n\n' :
-					renderValue.value) + ` ${ChatListItemRenderer.cursorCharacter}`;
-				const result = this.renderMarkdown(new MarkdownString(plusCursor), element, disposables, templateData, true);
-				// Doing the progressive render
-				dom.clearNode(templateData.value);
-				templateData.value.appendChild(result.element);
-				disposables.add(result);
+			} else if (!isFullyRendered) {
+				let hasRenderedOneMarkdownBlock = false;
+				partsToRender.forEach((partToRender, index) => {
+					if (!partToRender) {
+						return;
+					}
+
+					let result;
+					if ('label' in partToRender) {
```

Comment:
Would be nice to have a type predicate function for this since `'label' in xyz` comes up a lot and it's hard to remember which one that is